pkg/i2gw/providers/kong/converter.go

```diff
@@ -61,3 +66,29 @@ func (c *converter) ToGatewayAPI(resources i2gw.InputResources) (i2gw.GatewayRes
 
 	return gatewayResources, errs
 }
+
+func toHTTPRouteMatchOption(routePath networkingv1.HTTPIngressPath, path *field.Path) (*gatewayv1beta1.HTTPRouteMatch, *field.Error) {
+	pmPrefix := gatewayv1beta1.PathMatchPathPrefix
+	pmExact := gatewayv1beta1.PathMatchExact
+	pmRegex := gatewayv1beta1.PathMatchRegularExpression
+
+	match := &gatewayv1beta1.HTTPRouteMatch{Path: &gatewayv1beta1.HTTPPathMatch{Value: &routePath.Path}}
+	switch *routePath.PathType {
+	case networkingv1.PathTypePrefix:
+		match.Path.Type = &pmPrefix
+	case networkingv1.PathTypeExact:
+		match.Path.Type = &pmExact
+	case networkingv1.PathTypeImplementationSpecific:
+		if strings.HasPrefix(routePath.Path, "/~") {
+			match.Path.Type = &pmRegex
+			match.Path.Value = common.PtrTo(strings.TrimPrefix(*match.Path.Value, "/~"))
+		} else {
+			match.Path.Type = &pmPrefix
+		}
+
+	default:
+		return nil, field.Invalid(path.Child("pathType"), routePath.PathType, fmt.Sprintf("unsupported path match type: %s", *routePath.PathType))
+	}
+
+	return match, nil
+}
```

Comment:
I would prefer to create a new function for the `PathTypeImplementationSpecific` logic and pass that to the common converter `toHTTPRouteMatchOption` somehow instead of duplicating the entire function.  Given that multiple different providers will have different implementations of the `PathTypeImplementationSpecific`.

I know this is going to be a pretty significant change, but this is what I would have liked to see (Though I'm very interested in other opinions):
1. At the common package, create a new struct named `Converter` and attach the `ToGateway` function to him.
2. Each provider is going to construct his own common converter with provider-specific configuration, for example: `PathTypeImplementationSpecific` function
3. Have the `toHTTPRouteMatchOption` function call the configured `PathTypeImplementationSpecific` function in the case of `PathTypeImplementationSpecific`.

```golang
type Converter struct {
    PathTypeImplementationSpecificFunction func() ...
}

func (c *Converter) ToGateway() ...
```

I think this change will minimize code duplication and allow for easier extensibility in the future instead of passing a new parameter through `ToGateway` each time.

Again, I'm very interested in other opinions regarding this one

Reply:
> 1. why do we need a new `Converter` struct ? we can add this field on the `converter` interface and have the providers initialize it when they initialize the provider.

I don't think this is going to be the only time we're going to have to pass new fields to the `converter`. Creating a new struct from the beginning will make things smoother for the following times, IMO.



> 2. [ToGatewayAPI](https://github.com/kubernetes-sigs/ingress2gateway/blob/38c60b4f573a9e7a52eae580745584f6e6c0d3a7/pkg/i2gw/providers/kong/converter.go#L46) is already a method on the `converter` interface so I do not see a point to change the `ToGateway` as well. we can pass the implementation specific filed down the chain, similar to what Mattia has already did.

`ToGatewayAPI` is a provider-specific method. `ToGateway` is the common one, and it is simply a function, not a method. Again, it comes down to not passing down endless parameters down the chain

Please do let me know if it makes more sense now in your opinion @LiorLieberman @mlavacca

Reply:
I get @levikobi point on having a common structured way to customize such fields that will be helpful in the future as well. I also agree that passing a long list of parameters to a function is something we should try to avoid as much as possible, and since potentially any implementation-specific parameter could be customized, I think it makes sense to introduce a general enough mechanism to do so.

Reply:
I have implemented an options-based mechanism. This way it will be easy to add implementation-specific functionalities in the future, keeping the maintenance burden and code duplication very low. Let me know what you think @levikobi @LiorLieberman